project/src/main/webapp/script.js

```diff
@@ -304,13 +348,70 @@ function retrieveProductInfo() {
   });
 }
 
+// Retrieves and displays the optional images in a table.
+function retrieveOptionalImages(product) {
+  const imageUrls = product.imageUrls.slice(1);
+  const gcsUrls = product.gcsUrls.slice(1);
+  document.getElementById("optionalGcsUrls").value = JSON.stringify(gcsUrls);
+  document.getElementById("optionalImageUrls").value = JSON.stringify(imageUrls);
+  const tableBody = document.getElementById("imageTableBody");
+
+  // ImageUrls and GcsUrls are guaranteed to have the same length
+  var i;
+  for (i = 0; i < imageUrls.length; i++) {
+    const row = document.createElement("tr");
+    const imageCell = document.createElement("td");
+    const imageLink = document.createElement("a");
+    imageLink.href = imageUrls[i];
+    imageLink.target = "_blank"; // Open image in a new tab.
+    const image = document.createElement("img");
+    image.src = imageUrls[i];
+    image.style.maxHeight = "50px";
+    imageLink.appendChild(image);
+    imageCell.appendChild(imageLink);
+    const buttonCell = document.createElement("td");
+    const button = document.createElement("input");
+    button.type = "button"
+    button.setAttribute('onclick','deleteRow(this)');
+    button.value = "Delete";
+    buttonCell.appendChild(button);
+    const imageUrl = document.createElement("td");
+    imageUrl.innerText = imageUrls[i];
+    imageUrl.style.display = "none";
+    const gcsUrl = document.createElement("td");   
+    gcsUrl.innerText = gcsUrls[i];
+    gcsUrl.style.display = "none";
+    row.appendChild(imageCell);
+    row.appendChild(buttonCell);
+    row.appendChild(gcsUrl);
+    row.appendChild(imageUrl);
+    tableBody.appendChild(row);
+  }
+}
+
+function deleteRow(r) {
+  var i = r.parentNode.parentNode.rowIndex;
+  const table = document.getElementById("imageTable");
+  table.deleteRow(i);
+  // Loop through all the rows in the table and get the gcs and imageUrls.
+  var j;
+  const gcsUrls = [];
+  const imageUrls = [];
+  // First row ist just the headers of the columns, so we don't include that
```

Comment:
Good idea to put a comment here because it makes it clear why you're not doing normal behavior like starting the loop at 0.  Just minor spelling error: ist/is